File: code/java/MultiLayerPerceptron.java
/*
 * Multi-Layer Perceptron Implementation in Java
 * A neural network with multiple hidden layers for binary classification.
 */

import java.util.*;
import java.util.function.Function;

// Activation function enums
enum ActivationType {
    SIGMOID, RELU, TANH
}

// Matrix class
class Matrix {
    private double[][] data;
    private int rows;
    private int cols;
    
    public Matrix(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.data = new double[rows][cols];
    }
    
    public Matrix(double[][] data) {
        this.rows = data.length;
        this.cols = data[0].length;
        this.data = new double[rows][cols];
        for (int i = 0; i < rows; i++) {
            System.arraycopy(data[i], 0, this.data[i], 0, cols);
        }
    }
    
    public int getRows() { return rows; }
    public int getCols() { return cols; }
    public double[][] getData() { return data; }
    
    public double get(int i, int j) { return data[i][j]; }
    public void set(int i, int j, double value) { data[i][j] = value; }
    
    public Matrix multiply(Matrix other) {
        if (this.cols != other.rows) {
            throw new IllegalArgumentException("Matrix dimensions don't match for multiplication");
        }
        
        Matrix result = new Matrix(this.rows, other.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < other.cols; j++) {
                for (int k = 0; k < this.cols; k++) {
                    result.data[i][j] += this.data[i][k] * other.data[k][j];
                }
            }
        }
        return result;
    }
    
    public Matrix add(Matrix other) {
        Matrix result = new Matrix(this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                result.data[i][j] = this.data[i][j] + other.data[i][j];
            }
        }
        return result;
    }
    
    public Matrix subtract(Matrix other) {
        Matrix result = new Matrix(this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                result.data[i][j] = this.data[i][j] - other.data[i][j];
            }
        }
        return result;
    }
    
    public Matrix transpose() {
        Matrix result = new Matrix(this.cols, this.rows);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                result.data[j][i] = this.data[i][j];
            }
        }
        return result;
    }
    
    public Matrix apply(Function<Double, Double> func) {
        Matrix result = new Matrix(this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                result.data[i][j] = func.apply(this.data[i][j]);
            }
        }
        return result;
    }
    
    public Matrix elementwiseMultiply(Matrix other) {
        Matrix result = new Matrix(this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                result.data[i][j] = this.data[i][j] * other.data[i][j];
            }
        }
        return result;
    }
    
    public Matrix copy() {
        Matrix result = new Matrix(this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            System.arraycopy(this.data[i], 0, result.data[i], 0, this.cols);
        }
        return result;
    }
}

// Activation functions class
class Activations {
    public static double sigmoid(double z) {
        if (z < -500) z = -500;
        if (z > 500) z = 500;
        return 1.0 / (1.0 + Math.exp(-z));
    }
    
    public static double sigmoidDerivative(double a) {
        return a * (1.0 - a);
    }
    
    public static double relu(double z) {
        return Math.max(0, z);
    }
    
    public static double reluDerivative(double z) {
        return z > 0 ? 1.0 : 0.0;
    }
    
    public static double tanh(double z) {
        return Math.tanh(z);
    }
    
    public static double tanhDerivative(double a) {
        return 1.0 - a * a;
    }
}

// Layer class
class Layer {
    private Matrix W;
    private Matrix b;
    private Matrix z;
    private Matrix a;
    private Matrix x;
    private ActivationType activation;
    
    public Layer(int inputSize, int outputSize, ActivationType activation) {
        this.W = new Matrix(outputSize, inputSize);
        this.b = new Matrix(outputSize, 1);
        this.activation = activation;
        
        // Xavier/Glorot initialization
        double limit = Math.sqrt(6.0 / (inputSize + outputSize));
        Random rand = new Random();
        
        for (int i = 0; i < outputSize; i++) {
            for (int j = 0; j < inputSize; j++) {
                W.set(i, j, rand.nextDouble() * 2 * limit - limit);
            }
        }
    }
    
    public Matrix forward(Matrix input) {
        this.x = input.copy();
        
        // z = W * x + b
        this.z = W.multiply(input);
        for (int i = 0; i < z.getRows(); i++) {
            for (int j = 0; j < z.getCols(); j++) {
                z.set(i, j, z.get(i, j) + b.get(i, 0));
            }
        }
        
        // Apply activation
        switch (activation) {
            case SIGMOID:
                this.a = z.apply(Activations::sigmoid);
                break;
            case RELU:
                this.a = z.apply(Activations::relu);
                break;
            case TANH:
                this.a = z.apply(Activations::tanh);
                break;
        }
        
        return this.a;
    }
    
    public LayerGradients backward(Matrix deltaNext) {
        int batchSize = x.getCols();
        
        // Apply activation derivative
        Matrix delta;
        if (activation == ActivationType.SIGMOID || activation == ActivationType.TANH) {
            Function<Double, Double> derivFunc = (activation == ActivationType.SIGMOID) 
                ? Activations::sigmoidDerivative 
                : Activations::tanhDerivative;
            
            Matrix deriv = a.apply(derivFunc);
            delta = deltaNext.elementwiseMultiply(deriv);
        } else { // RELU
            Matrix deriv = z.apply(Activations::reluDerivative);
            delta = deltaNext.elementwiseMultiply(deriv);
        }
        
        // Compute gradients
        Matrix dW = delta.multiply(x.transpose());
        for (int i = 0; i < dW.getRows(); i++) {
            for (int j = 0; j < dW.getCols(); j++) {
                dW.set(i, j, dW.get(i, j) / batchSize);
            }
        }
        
        Matrix db = new Matrix(delta.getRows(), 1);
        for (int i = 0; i < delta.getRows(); i++) {
            double sum = 0;
            for (int j = 0; j < delta.getCols(); j++) {
                sum += delta.get(i, j);
            }
            db.set(i, 0, sum / batchSize);
        }
        
        // Gradient for previous layer
        Matrix deltaPrev = W.transpose().multiply(delta);
        
        return new LayerGradients(deltaPrev, dW, db);
    }
    
    public Matrix getW() { return W; }
    public Matrix getB() { return b; }
    public Matrix getX() { return x; }
    public void updateW(Matrix dW, double learningRate) {
        for (int i = 0; i < W.getRows(); i++) {
            for (int j = 0; j < W.getCols(); j++) {
                W.set(i, j, W.get(i, j) - learningRate * dW.get(i, j));
            }
        }
    }
    
    public void updateB(Matrix db, double learningRate) {
        for (int i = 0; i < b.getRows(); i++) {
            b.set(i, 0, b.get(i, 0) - learningRate * db.get(i, 0));
        }
    }
}

// Helper class for layer gradients
class LayerGradients {
    Matrix deltaPrev;
    Matrix dW;
    Matrix db;
    
    public LayerGradients(Matrix deltaPrev, Matrix dW, Matrix db) {
        this.deltaPrev = deltaPrev;
        this.dW = dW;
        this.db = db;
    }
}

// MLP class
class MLP {
    private List<Layer> layers;
    private List<Double> losses;
    
    public MLP(int[] layerSizes, ActivationType[] activations) {
        this.layers = new ArrayList<>();
        this.losses = new ArrayList<>();
        
        for (int i = 0; i < layerSizes.length - 1; i++) {
            layers.add(new Layer(layerSizes[i], layerSizes[i+1], activations[i]));
        }
    }
    
    public Matrix forward(Matrix X) {
        Matrix a = X;
        for (Layer layer : layers) {
            a = layer.forward(a);
        }
        return a;
    }
    
    public double binaryCrossEntropy(Matrix yTrue, Matrix yPred) {
        double loss = 0.0;
        double epsilon = 1e-15;
        int n = yTrue.getCols();
        
        for (int j = 0; j < n; j++) {
            double pred = Math.max(epsilon, Math.min(1.0 - epsilon, yPred.get(0, j)));
            double trueVal = yTrue.get(0, j);
            loss += trueVal * Math.log(pred) + (1.0 - trueVal) * Math.log(1.0 - pred);
        }
        
        return -loss / n;
    }
    
    public List<LayerGradients> backward(Matrix yTrue, Matrix yPred) {
        // Initial gradient
        Matrix delta = yPred.subtract(yTrue);
        
        List<LayerGradients> gradients = new ArrayList<>();
        
        // Backpropagate
        for (int i = layers.size() - 1; i >= 0; i--) {
            Layer layer = layers.get(i);
            LayerGradients grad;
            
            if (i == layers.size() - 1) {
                // Output layer
                int batchSize = layer.getX().getCols();
                Matrix dW = delta.multiply(layer.getX().transpose());
                for (int r = 0; r < dW.getRows(); r++) {
                    for (int c = 0; c < dW.getCols(); c++) {
                        dW.set(r, c, dW.get(r, c) / batchSize);
                    }
                }
                
                Matrix db = new Matrix(delta.getRows(), 1);
                for (int r = 0; r < delta.getRows(); r++) {
                    double sum = 0;
                    for (int c = 0; c < delta.getCols(); c++) {
                        sum += delta.get(r, c);
                    }
                    db.set(r, 0, sum / batchSize);
                }
                
                Matrix deltaPrev = layer.getW().transpose().multiply(delta);
                grad = new LayerGradients(deltaPrev, dW, db);
            } else {
                // Hidden layers
                grad = layer.backward(delta);
            }
            
            gradients.add(0, grad);
            delta = grad.deltaPrev;
        }
        
        return gradients;
    }
    
    public void updateParameters(List<LayerGradients> gradients, double learningRate) {
        for (int i = 0; i < layers.size(); i++) {
            Layer layer = layers.get(i);
            LayerGradients grad = gradients.get(i);
            layer.updateW(grad.dW, learningRate);
            layer.updateB(grad.db, learningRate);
        }
    }
    
    public void fit(Matrix X, double[] y, double learningRate, int epochs, boolean verbose) {
        // Transpose input
        Matrix XT = X.transpose();
        Matrix yT = new Matrix(1, y.length);
        for (int i = 0; i < y.length; i++) {
            yT.set(0, i, y[i]);
        }
        
        losses.clear();
        
        for (int epoch = 0; epoch < epochs; epoch++) {
            // Forward pass
            Matrix yPred = forward(XT);
            
            // Compute loss
            double loss = binaryCrossEntropy(yT, yPred);
            losses.add(loss);
            
            // Backward pass
            List<LayerGradients> gradients = backward(yT, yPred);
            
            // Update parameters
            updateParameters(gradients, learningRate);
            
            // Print progress
            if (verbose && (epoch + 1) % 1000 == 0) {
                System.out.printf("Epoch %d/%d, Loss: %.6f%n", epoch + 1, epochs, loss);
            }
        }
        
        if (verbose) {
            System.out.printf("%nTraining completed. Final loss: %.6f%n", losses.get(losses.size() - 1));
        }
    }
    
    public Matrix predictProba(Matrix X) {
        Matrix XT = X.transpose();
        Matrix result = forward(XT);
        return result.transpose();
    }
    
    public int[] predict(Matrix X, double threshold) {
        Matrix proba = predictProba(X);
        int[] predictions = new int[proba.getRows()];
        
        for (int i = 0; i < proba.getRows(); i++) {
            predictions[i] = proba.get(i, 0) >= threshold ? 1 : 0;
        }
        
        return predictions;
    }
    
    public double score(Matrix X, double[] y) {
        int[] predictions = predict(X, 0.5);
        int correct = 0;
        
        for (int i = 0; i < y.length; i++) {
            if (predictions[i] == (int)y[i]) {
                correct++;
            }
        }
        
        return (double)correct / y.length;
    }
}

// Main class
public class MultiLayerPerceptron {
    public static void main(String[] args) {
        System.out.println("Training MLP on XOR problem\n");
        
        // XOR dataset
        double[][] XData = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
        Matrix X = new Matrix(XData);
        double[] y = {0, 1, 1, 0};
        
        // Create and train MLP
        int[] layerSizes = {2, 4, 1};
        ActivationType[] activations = {ActivationType.TANH, ActivationType.SIGMOID};
        
        MLP mlp = new MLP(layerSizes, activations);
        mlp.fit(X, y, 0.5, 5000, true);
        
        // Test predictions
        System.out.println("\n--- Predictions ---");
        int[] predictions = mlp.predict(X, 0.5);
        Matrix probas = mlp.predictProba(X);
        
        for (int i = 0; i < 4; i++) {
            System.out.printf("Input: [%.0f, %.0f], Predicted: %d, Probability: %.4f, Actual: %.0f%n",
                XData[i][0], XData[i][1], predictions[i], probas.get(i, 0), y[i]);
        }
        
        System.out.printf("%nAccuracy: %.0f%%%n", mlp.score(X, y) * 100);
    }
}

File: code/go/multi_layer_perceptron.go
/*
 * Multi-Layer Perceptron Implementation in Go
 * A neural network with multiple hidden layers for binary classification.
 */

package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

// ActivationType represents different activation functions
type ActivationType int

const (
	Sigmoid ActivationType = iota
	ReLU
	Tanh
)

// Matrix represents a 2D matrix
type Matrix struct {
	Rows int
	Cols int
	Data [][]float64
}

// NewMatrix creates a new matrix with given dimensions
func NewMatrix(rows, cols int) *Matrix {
	data := make([][]float64, rows)
	for i := range data {
		data[i] = make([]float64, cols)
	}
	return &Matrix{Rows: rows, Cols: cols, Data: data}
}

// Multiply performs matrix multiplication
func (m *Matrix) Multiply(other *Matrix) *Matrix {
	result := NewMatrix(m.Rows, other.Cols)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < other.Cols; j++ {
			for k := 0; k < m.Cols; k++ {
				result.Data[i][j] += m.Data[i][k] * other.Data[k][j]
			}
		}
	}
	return result
}

// Add performs element-wise addition
func (m *Matrix) Add(other *Matrix) *Matrix {
	result := NewMatrix(m.Rows, m.Cols)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < m.Cols; j++ {
			result.Data[i][j] = m.Data[i][j] + other.Data[i][j]
		}
	}
	return result
}

// Subtract performs element-wise subtraction
func (m *Matrix) Subtract(other *Matrix) *Matrix {
	result := NewMatrix(m.Rows, m.Cols)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < m.Cols; j++ {
			result.Data[i][j] = m.Data[i][j] - other.Data[i][j]
		}
	}
	return result
}

// Transpose returns the transpose of the matrix
func (m *Matrix) Transpose() *Matrix {
	result := NewMatrix(m.Cols, m.Rows)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < m.Cols; j++ {
			result.Data[j][i] = m.Data[i][j]
		}
	}
	return result
}

// Apply applies a function element-wise
func (m *Matrix) Apply(fn func(float64) float64) *Matrix {
	result := NewMatrix(m.Rows, m.Cols)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < m.Cols; j++ {
			result.Data[i][j] = fn(m.Data[i][j])
		}
	}
	return result
}

// ElementwiseMultiply performs element-wise multiplication
func (m *Matrix) ElementwiseMultiply(other *Matrix) *Matrix {
	result := NewMatrix(m.Rows, m.Cols)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < m.Cols; j++ {
			result.Data[i][j] = m.Data[i][j] * other.Data[i][j]
		}
	}
	return result
}

// Copy creates a copy of the matrix
func (m *Matrix) Copy() *Matrix {
	result := NewMatrix(m.Rows, m.Cols)
	for i := 0; i < m.Rows; i++ {
		copy(result.Data[i], m.Data[i])
	}
	return result
}

// Activation functions
func sigmoid(z float64) float64 {
	if z < -500 {
		z = -500
	}
	if z > 500 {
		z = 500
	}
	return 1.0 / (1.0 + math.Exp(-z))
}

func sigmoidDerivative(a float64) float64 {
	return a * (1.0 - a)
}

func relu(z float64) float64 {
	return math.Max(0, z)
}

func reluDerivative(z float64) float64 {
	if z > 0 {
		return 1.0
	}
	return 0.0
}

func tanhFunc(z float64) float64 {
	return math.Tanh(z)
}

func tanhDerivative(a float64) float64 {
	return 1.0 - a*a
}

// Layer represents a single fully-connected layer
type Layer struct {
	W          *Matrix
	B          *Matrix
	Z          *Matrix
	A          *Matrix
	X          *Matrix
	Activation ActivationType
}

// NewLayer creates a new layer with given dimensions
func NewLayer(inputSize, outputSize int, activation ActivationType) *Layer {
	layer := &Layer{
		W:          NewMatrix(outputSize, inputSize),
		B:          NewMatrix(outputSize, 1),
		Activation: activation,
	}

	// Xavier/Glorot initialization
	limit := math.Sqrt(6.0 / float64(inputSize+outputSize))
	rand.Seed(time.Now().UnixNano())

	for i := 0; i < outputSize; i++ {
		for j := 0; j < inputSize; j++ {
			layer.W.Data[i][j] = rand.Float64()*2*limit - limit
		}
	}

	return layer
}

// Forward performs forward pass through the layer
func (l *Layer) Forward(input *Matrix) *Matrix {
	l.X = input.Copy()

	// z = W * x + b
	l.Z = l.W.Multiply(input)
	for i := 0; i < l.Z.Rows; i++ {
		for j := 0; j < l.Z.Cols; j++ {
			l.Z.Data[i][j] += l.B.Data[i][0]
		}
	}

	// Apply activation
	switch l.Activation {
	case Sigmoid:
		l.A = l.Z.Apply(sigmoid)
	case ReLU:
		l.A = l.Z.Apply(relu)
	case Tanh:
		l.A = l.Z.Apply(tanhFunc)
	}

	return l.A
}

// Backward performs backward pass through the layer
func (l *Layer) Backward(deltaNext *Matrix) (*Matrix, *Matrix, *Matrix) {
	batchSize := float64(l.X.Cols)

	// Apply activation derivative
	var delta *Matrix
	if l.Activation == Sigmoid || l.Activation == Tanh {
		var derivFn func(float64) float64
		if l.Activation == Sigmoid {
			derivFn = sigmoidDerivative
		} else {
			derivFn = tanhDerivative
		}
		deriv := l.A.Apply(derivFn)
		delta = deltaNext.ElementwiseMultiply(deriv)
	} else { // ReLU
		deriv := l.Z.Apply(reluDerivative)
		delta = deltaNext.ElementwiseMultiply(deriv)
	}

	// Compute gradients
	dW := delta.Multiply(l.X.Transpose())
	for i := 0; i < dW.Rows; i++ {
		for j := 0; j < dW.Cols; j++ {
			dW.Data[i][j] /= batchSize
		}
	}

	dB := NewMatrix(delta.Rows, 1)
	for i := 0; i < delta.Rows; i++ {
		sum := 0.0
		for j := 0; j < delta.Cols; j++ {
			sum += delta.Data[i][j]
		}
		dB.Data[i][0] = sum / batchSize
	}

	// Gradient for previous layer
	deltaPrev := l.W.Transpose().Multiply(delta)

	return deltaPrev, dW, dB
}

// MLP represents a multi-layer perceptron
type MLP struct {
	Layers []Layer
	Losses []float64
}

// NewMLP creates a new MLP with given architecture
func NewMLP(layerSizes []int, activations []ActivationType) *MLP {
	mlp := &MLP{
		Layers: make([]Layer, len(layerSizes)-1),
		Losses: make([]float64, 0),
	}

	for i := 0; i < len(layerSizes)-1; i++ {
		mlp.Layers[i] = *NewLayer(layerSizes[i], layerSizes[i+1], activations[i])
	}

	return mlp
}

// Forward performs forward pass through the MLP
func (mlp *MLP) Forward(X *Matrix) *Matrix {
	a := X
	for i := range mlp.Layers {
		a = mlp.Layers[i].Forward(a)
	}
	return a
}

// BinaryCrossEntropy computes binary cross-entropy loss
func BinaryCrossEntropy(yTrue, yPred *Matrix) float64 {
	loss := 0.0
	epsilon := 1e-15
	n := float64(yTrue.Cols)

	for j := 0; j < yTrue.Cols; j++ {
		pred := yPred.Data[0][j]
		pred = math.Max(epsilon, math.Min(1.0-epsilon, pred))
		trueVal := yTrue.Data[0][j]
		loss += trueVal*math.Log(pred) + (1.0-trueVal)*math.Log(1.0-pred)
	}

	return -loss / n
}

// Backward performs backward pass through the MLP
func (mlp *MLP) Backward(yTrue, yPred *Matrix) [][2]*Matrix {
	// Initial gradient
	delta := yPred.Subtract(yTrue)

	gradients := make([][2]*Matrix, len(mlp.Layers))

	// Backpropagate
	for i := len(mlp.Layers) - 1; i >= 0; i-- {
		var deltaPrev, dW, dB *Matrix

		if i == len(mlp.Layers)-1 {
			// Output layer
			batchSize := float64(mlp.Layers[i].X.Cols)
			dW = delta.Multiply(mlp.Layers[i].X.Transpose())
			for r := 0; r < dW.Rows; r++ {
				for c := 0; c < dW.Cols; c++ {
					dW.Data[r][c] /= batchSize
				}
			}

			dB = NewMatrix(delta.Rows, 1)
			for r := 0; r < delta.Rows; r++ {
				sum := 0.0
				for c := 0; c < delta.Cols; c++ {
					sum += delta.Data[r][c]
				}
				dB.Data[r][0] = sum / batchSize
			}

			deltaPrev = mlp.Layers[i].W.Transpose().Multiply(delta)
		} else {
			// Hidden layers
			deltaPrev, dW, dB = mlp.Layers[i].Backward(delta)
		}

		gradients[i] = [2]*Matrix{dW, dB}
		delta = deltaPrev
	}

	return gradients
}

// UpdateParameters updates layer parameters using gradients
func (mlp *MLP) UpdateParameters(gradients [][2]*Matrix, learningRate float64) {
	for i := range mlp.Layers {
		dW := gradients[i][0]
		dB := gradients[i][1]

		for r := 0; r < mlp.Layers[i].W.Rows; r++ {
			for c := 0; c < mlp.Layers[i].W.Cols; c++ {
				mlp.Layers[i].W.Data[r][c] -= learningRate * dW.Data[r][c]
			}
		}

		for r := 0; r < mlp.Layers[i].B.Rows; r++ {
			mlp.Layers[i].B.Data[r][0] -= learningRate * dB.Data[r][0]
		}
	}
}

// Fit trains the MLP
func (mlp *MLP) Fit(X *Matrix, y []float64, learningRate float64, epochs int, verbose bool) {
	// Transpose input
	XT := X.Transpose()
	yT := NewMatrix(1, len(y))
	for i := range y {
		yT.Data[0][i] = y[i]
	}

	mlp.Losses = make([]float64, 0)

	for epoch := 0; epoch < epochs; epoch++ {
		// Forward pass
		yPred := mlp.Forward(XT)

		// Compute loss
		loss := BinaryCrossEntropy(yT, yPred)
		mlp.Losses = append(mlp.Losses, loss)

		// Backward pass
		gradients := mlp.Backward(yT, yPred)

		// Update parameters
		mlp.UpdateParameters(gradients, learningRate)

		// Print progress
		if verbose && (epoch+1)%1000 == 0 {
			fmt.Printf("Epoch %d/%d, Loss: %.6f\n", epoch+1, epochs, loss)
		}
	}

	if verbose {
		fmt.Printf("\nTraining completed. Final loss: %.6f\n", mlp.Losses[len(mlp.Losses)-1])
	}
}

// PredictProba predicts probabilities
func (mlp *MLP) PredictProba(X *Matrix) *Matrix {
	XT := X.Transpose()
	result := mlp.Forward(XT)
	return result.Transpose()
}

// Predict predicts class labels
func (mlp *MLP) Predict(X *Matrix, threshold float64) []int {
	proba := mlp.PredictProba(X)
	predictions := make([]int, proba.Rows)

	for i := 0; i < proba.Rows; i++ {
		if proba.Data[i][0] >= threshold {
			predictions[i] = 1
		} else {
			predictions[i] = 0
		}
	}

	return predictions
}

// Score computes accuracy
func (mlp *MLP) Score(X *Matrix, y []float64) float64 {
	predictions := mlp.Predict(X, 0.5)
	correct := 0

	for i := range y {
		if predictions[i] == int(y[i]) {
			correct++
		}
	}

	return float64(correct) / float64(len(y))
}

func main() {
	fmt.Println("Training MLP on XOR problem\n")

	// XOR dataset
	X := NewMatrix(4, 2)
	X.Data = [][]float64{
		{0, 0},
		{0, 1},
		{1, 0},
		{1, 1},
	}

	y := []float64{0, 1, 1, 0}

	// Create and train MLP
	mlp := NewMLP([]int{2, 4, 1}, []ActivationType{Tanh, Sigmoid})
	mlp.Fit(X, y, 0.5, 5000, true)

	// Test predictions
	fmt.Println("\n--- Predictions ---")
	predictions := mlp.Predict(X, 0.5)
	probas := mlp.PredictProba(X)

	for i := 0; i < 4; i++ {
		fmt.Printf("Input: [%.0f, %.0f], Predicted: %d, Probability: %.4f, Actual: %.0f\n",
			X.Data[i][0], X.Data[i][1], predictions[i], probas.Data[i][0], y[i])
	}

	fmt.Printf("\nAccuracy: %.0f%%\n", mlp.Score(X, y)*100)
}
